src/EventStore.Core.XUnit.Tests/Scavenge/CollisionDetectorTests.cs

```diff
@@ -0,0 +1,123 @@
+				("b-stream2", none),
+				("a-streamOfInterest", none));
+
+			yield return Case(
+				"seen stream before. was not a collision but has since been collided with",
+				("a-streamOfInterest", none),
+				("a-stream1", new[] { "a-stream1", "a-streamOfInterest" }),
+				("b-stream2", none),
+				("a-streamOfInterest", none));
+
+			yield return Case(
+				"seen stream before. was not a collision and still isnt",
+				("a-streamOfInterest", none),
+				("b-stream2", none),
+				("a-streamOfInterest", none));
+
+			yield return Case(
+				"first time seeing stream. no collision",
+				("a-stream1", none),
+				("b-stream2", none),
+				("c-streamOfInterest", none));
+
+			yield return Case(
+				"first time seeing stream. collides with previous stream",
+				("a-stream1", none),
+				("b-stream2", none),
+				("a-streamOfInterest", new[] { "a-stream1", "a-streamOfInterest" }));
+
+			yield return Case(
+				"three way collision",
+				("a-stream1", none),
+				("a-stream2", new[] { "a-stream1", "a-stream2" }),
+				("a-stream3", new[] { "a-stream3" }));
+
+			yield return Case(
+				"in combination",
+				("a-stream1", none), // 2b
+				("a-stream2", new[] { "a-stream1", "a-stream2" }), // 2a
+				("b-stream3", none),
+				("a-stream4", new[] { "a-stream4" }),
+				("b-stream3", none), // 1c
+				("a-stream1", none), // 1b
+				("a-stream2", none)); // 1a
+
+			object[] Case(string name, params (string, string[])[] data) {
+				return new object[] {
+					name, data
+				};
+			}
+		}
+
+		[Theory]
+		[MemberData(nameof(TheCases))]
+		public void Works(string caseName, (string StreamName, string[] NewCollisions)[] data) {
+			Assert.NotNull(caseName);
+
+			var log = data.Select(x => x.StreamName).ToArray();
+			var hasher = new FirstCharacterHasher();
+
+			// index maps hashes to lists of log positions
+			var index = new Dictionary<ulong, List<int>>();
+
+			// populate the index
+			for (var i = 0; i < log.Length; i++) {
+				var streamName = log[i];
+				var hash = hasher.Hash(streamName);
+				if (!index.TryGetValue(hash, out var entries)) {
+					entries = new List<int>();
+					index[hash] = entries;
+				}
+
+				// which order is more realistic? this way shows the need for filtering by position anyway
+				entries.Insert(0, i);
+			}
+
+			var hashes = new SqliteScavengeMap<ulong, string>("hashes");
+			hashes.Initialize(new SqliteBackend(Fixture.DbConnection));
+
+			var collisions = new SqliteCollisionScavengeMap<string>();
+			collisions.Initialize(new SqliteBackend(Fixture.DbConnection));
+
+			var sut = new CollisionDetector<string>(
+				new LruCachingScavengeMap<ulong, string>(
+					hashes,
+					cacheMaxCount: 1000),
+				collisions,
+				hasher);
+
+			var expectedCollisions = new HashSet<string>();
+
+			for (var i = 0; i < data.Length; i++) {
+				foreach (var newCollision in data[i].NewCollisions)
+					expectedCollisions.Add(newCollision);
+
+				sut.DetectCollisions(data[i].StreamName, out _);
```

Comment:
suggestion: we could add the expected CollisionResult to the data tuple and verify it here to further strengthen the test

Reply:
:+1: done